library/core/src/hint.rs

```diff
@@ -382,6 +391,85 @@ pub fn spin_loop() {
 ///
 /// This makes our benchmark much more realistic to how the function would actually be used, where
 /// arguments are usually not known at compile time and the result is used in some way.
+///
+/// # How to use this
+///
+/// In practice, `black_box` serves two purposes:
+///
+/// 1. It prevents the compiler from making optimizations related to the value of the returned
+///    type
+/// 2. It forces the input to be calculated, even if its results are never used
+///
+/// ```
+/// use std::hint::black_box;
+///
+/// let zero = 0;
+/// let five = 5;
+///
+/// // The compiler will see this and remove the `* five` call, because it knows that multiplying
+/// // any integer by 0 will result in 0. This is a value optimization: the compiler knows the
+/// // value of `zero` must be 0, and thus can make optimizations related to that.
+/// let c = zero * five;
+///
+/// // Adding `black_box` here disables the compiler's ability to reason about the value of `zero`.
+/// // It is forced to assume that it can be any possible number, so it cannot remove the `* five`
+/// // operation.
```

Comment:
```suggestion
/// // Adding `black_box` here disables the compiler's ability to reason about the first operand in the multiplication.
/// // It is forced to assume that it can be any possible number, so it cannot remove the `* five`
/// // operation.
```

Your initial wording here is wrong, the compiler is still perfectly capable of reasoning about the value of `zero`, what it cannot reason about is the value of the left-hand-side operand to `*` in this expression. Later mentions of `zero` will be just as optimizable. If you wanted to inhibit optimizations of `zero` based on the value it is initialized with, you would create `zero` with `let zero = black_box(0);`, which is a pattern I've seen and used myself.